 # To view NYC Crime 2020 dashboard please go to: 

https://public.tableau.com/views/NYCCrime2020Dashboard/StoryofNYCCrimein2020?:language=en-US&publish=yes&:sid=&:redirect=auth&:display_count=n&:origin=viz_share_link

# NYC Crime in 2020 Dataset Cleanup

Four datasets were combined and cleaned to make a complete crime record for NYC Neighborhoods 2020.

NYPD Complaint Data Link (updated May 3, 2021): https://data.cityofnewyork.us/Public-Safety/NYPD-Complaint-Data-Historic/qgea-i56i

NYPD Shooting Data Link (updated May 3, 2021): https://data.cityofnewyork.us/Public-Safety/NYPD-Shooting-Incident-Data-Historic-/833y-fsy8

MODZCTA Link (updated May 19, 2020): https://data.cityofnewyork.us/Health/Modified-Zip-Code-Tabulation-Areas-MODZCTA-/pri4-ifjk

Neighborhood Data Link: https://raw.githubusercontent.com/nychealth/coronavirus-data/master/latest/last7days-by-modzcta.csv

### Import Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.neighbors import KDTree

import warnings
warnings.filterwarnings("ignore")


### Import NYPD Complaint Data

In [2]:
df_crime = pd.read_csv("NYPD_Complaint_Data_Historic.csv")

In [3]:
df_crime.head()

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM  \
0   700381962   05/28/2015     15:00:00          NaN          NaN   
1   642234217   10/28/2013     13:50:00   10/28/2013     13:50:00   
2   242465164   05/09/2012     20:50:00   05/09/2012     21:00:00   
3   927207428   01/03/2014     13:30:00   01/03/2014     13:35:00   
4   492142357   04/13/2016     00:00:00          NaN          NaN   

   ADDR_PCT_CD      RPT_DT  KY_CD                       OFNS_DESC  PD_CD  ...  \
0         46.0  06/01/2015    578                   HARRASSMENT 2  638.0  ...   
1        120.0  10/28/2013    351  CRIMINAL MISCHIEF & RELATED OF  259.0  ...   
2         24.0  05/09/2012    236               DANGEROUS WEAPONS  782.0  ...   
3        108.0  01/03/2014    109                   GRAND LARCENY  409.0  ...   
4         40.0  04/13/2016    351  CRIMINAL MISCHIEF & RELATED OF  258.0  ...   

  SUSP_SEX TRANSIT_DISTRICT   Latitude  Longitude  \
0        M              NaN  40.845868 -73.915888   
1      NaN              NaN  40.627061 -74.077149   
2      NaN              NaN  40.800966 -73.969047   
3        M              NaN  40.745242 -73.894253   
4      NaN              NaN  40.810352 -73.924942   

                         Lat_Lon                PATROL_BORO STATION_NAME  \
0   (40.84586773, -73.915888033)          PATROL BORO BRONX          NaN   
1  (40.627060894, -74.077149232)  PATROL BORO STATEN ISLAND          NaN   
2  (40.800965968, -73.969047272)      PATROL BORO MAN NORTH          NaN   
3  (40.745241809, -73.894253382)   PATROL BORO QUEENS NORTH          NaN   
4  (40.810351863, -73.924942326)          PATROL BORO BRONX          NaN   

   VIC_AGE_GROUP                  VIC_RACE VIC_SEX  
0          25-44            WHITE HISPANIC       F  
1          45-64                     WHITE       M  
2            NaN                   UNKNOWN       E  
3          45-64  ASIAN / PACIFIC ISLANDER       M  
4        UNKNOWN                   UNKNOWN       E  

[5 rows x 35 columns]

In [4]:
df_crime.shape

(7396619, 35)

Before cleanup there are 7396619 rows and 35 columns.

### Missing Values

In [5]:
df_crime.isnull().sum()

CMPLNT_NUM                 0
CMPLNT_FR_DT             655
CMPLNT_FR_TM              48
CMPLNT_TO_DT         1707480
CMPLNT_TO_TM         1702808
ADDR_PCT_CD             2166
RPT_DT                     0
KY_CD                      0
OFNS_DESC              18831
PD_CD                   6328
PD_DESC                 6328
CRM_ATPT_CPTD_CD           7
LAW_CAT_CD                 0
BORO_NM                11379
LOC_OF_OCCUR_DESC    1546078
PREM_TYP_DESC          40904
JURIS_DESC                 0
JURISDICTION_CODE       6328
PARKS_NM             7368895
HADEVELOPT           7049021
HOUSING_PSA          6828743
X_COORD_CD             24064
Y_COORD_CD             24064
SUSP_AGE_GROUP       4802435
SUSP_RACE            3433894
SUSP_SEX             3567208
TRANSIT_DISTRICT     7232302
Latitude               24064
Longitude              24064
Lat_Lon                24064
PATROL_BORO             6785
STATION_NAME         7232302
VIC_AGE_GROUP        1638445
VIC_RACE                 309
VIC_SEX       

The columns that we want to remove nulls are as follows:
- BORO_NM: borough name
- CMPLNT_FR_DT: date of crime occurence
- Latitude/Longitude: the latitude and longitude of the crime 

In [6]:
#subsetting rows with no nulls in BORO_NM
df_crime = df_crime[pd.notnull(df_crime['BORO_NM'])]

In [7]:
#subsetting rows with no nulls in CMPLNT_FR_DT
df_crime = df_crime[pd.notnull(df_crime['CMPLNT_FR_DT'])]

In [8]:
#subsetting rows with no nulls in Lat_Lon
df_crime = df_crime[pd.notnull(df_crime['Lat_Lon'])]

In [9]:
#checking nulls again
df_crime.isnull().sum()

CMPLNT_NUM                 0
CMPLNT_FR_DT               0
CMPLNT_FR_TM              47
CMPLNT_TO_DT         1692459
CMPLNT_TO_TM         1687839
ADDR_PCT_CD                2
RPT_DT                     0
KY_CD                      0
OFNS_DESC              18684
PD_CD                      0
PD_DESC                    0
CRM_ATPT_CPTD_CD           7
LAW_CAT_CD                 0
BORO_NM                    0
LOC_OF_OCCUR_DESC    1540211
PREM_TYP_DESC          33949
JURIS_DESC                 0
JURISDICTION_CODE          0
PARKS_NM             7333302
HADEVELOPT           7015031
HOUSING_PSA          6794904
X_COORD_CD                 0
Y_COORD_CD                 0
SUSP_AGE_GROUP       4778903
SUSP_RACE            3415835
SUSP_SEX             3548076
TRANSIT_DISTRICT     7197485
Latitude                   0
Longitude                  0
Lat_Lon                    0
PATROL_BORO               78
STATION_NAME         7197485
VIC_AGE_GROUP        1631632
VIC_RACE                 306
VIC_SEX       

There are now no null values for the columns that matter to us. 

### Filter Date Range

The dataset is supposed only include dates from 2006 until the present. Furthermore, in order to work with a smaller dataset that is easier to process, we will choose the date range 2018 to 2020. 

First we need to convert CMPLNT_FR_DT to datetime format. 

In [10]:
#convert CMPLNT_FR_DT column to datetime, 'coerce' forces dates that are out-of-bounds to be null values
#out-of-bounds means dates are not within 584 years 

df_crime['CMPLNT_FR_DT'] = pd.to_datetime(df_crime['CMPLNT_FR_DT'], errors = 'coerce')

In [11]:
#filter for dates that are 2020-01-01 until the present 
df_crime = df_crime[df_crime['CMPLNT_FR_DT'] >= '2020-01-01']

In [12]:
#check the minimum date to make sure filtering worked
df_crime['CMPLNT_FR_DT'].min()

Timestamp('2020-01-01 00:00:00')

Now, df_crime has the correct date range (2020 - present). 

### Drop Duplicates

In [13]:
#how many rows there are
df_crime.shape

(404429, 35)

In [14]:
#drop duplicates
df_crime.drop_duplicates(subset = ['CMPLNT_NUM'], inplace = True)

In [15]:
#how many row there are without duplicates
df_crime.shape

(404429, 35)

The total number of instances of each identifier is the same as the number of rows in the dataframe so there are no duplicates.

### Import NYPD Shooting Data

In [16]:
df_shooting = pd.read_csv("NYPD_Shooting_Incident_Data__Historic_.csv")

In [17]:
df_shooting.head()

INCIDENT_KEY  OCCUR_DATE OCCUR_TIME           BORO  PRECINCT  \
0     201575314  08/23/2019   22:10:00         QUEENS       103   
1     205748546  11/27/2019   15:54:00          BRONX        40   
2     193118596  02/02/2019   19:40:00      MANHATTAN        23   
3     204192600  10/24/2019   00:52:00  STATEN ISLAND       121   
4     201483468  08/22/2019   18:03:00          BRONX        46   

   JURISDICTION_CODE LOCATION_DESC  STATISTICAL_MURDER_FLAG PERP_AGE_GROUP  \
0                0.0           NaN                    False            NaN   
1                0.0           NaN                    False            <18   
2                0.0           NaN                    False          18-24   
3                0.0     PVT HOUSE                     True          25-44   
4                0.0           NaN                    False          25-44   

  PERP_SEX       PERP_RACE VIC_AGE_GROUP VIC_SEX        VIC_RACE X_COORD_CD  \
0      NaN             NaN         25-44       M           BLACK    1037451   
1        M           BLACK         25-44       F           BLACK    1006789   
2        M  WHITE HISPANIC         18-24       M  BLACK HISPANIC     999347   
3        M           BLACK         25-44       F           BLACK     938149   
4        M  BLACK HISPANIC         18-24       M           BLACK    1008224   

  Y_COORD_CD   Latitude  Longitude  \
0     193561  40.697805 -73.808141   
1     237559  40.818700 -73.918571   
2     227795  40.791916 -73.945480   
3     171781  40.638064 -74.166108   
4     250621  40.854547 -73.913339   

                                         Lon_Lat  
0  POINT (-73.80814071699996 40.697805308000056)  
1   POINT (-73.91857061799993 40.81869973000005)  
2  POINT (-73.94547965999999 40.791916091000076)  
3   POINT (-74.16610830199996 40.63806398200006)  
4   POINT (-73.91333944399999 40.85454734900003)

In [18]:
df_shooting.shape

(23568, 19)

Before cleanup there are 23,569 rows and 19 columns.

### Missing Data

In [19]:
df_shooting.isnull().sum()

INCIDENT_KEY                   0
OCCUR_DATE                     0
OCCUR_TIME                     0
BORO                           0
PRECINCT                       0
JURISDICTION_CODE              2
LOCATION_DESC              13581
STATISTICAL_MURDER_FLAG        0
PERP_AGE_GROUP              8459
PERP_SEX                    8425
PERP_RACE                   8425
VIC_AGE_GROUP                  0
VIC_SEX                        0
VIC_RACE                       0
X_COORD_CD                     0
Y_COORD_CD                     0
Latitude                       0
Longitude                      0
Lon_Lat                        0
dtype: int64

There is no missing information in the important columns for our analysis. 

### Filter Date Range

We will filter for the year 2020. 

In [20]:
#convert OCCUR_DATE column to datetime, 'coerce' forces dates that are out-of-bounds to be null values
#out-of-bounds means dates are not within 584 years 

df_shooting['OCCUR_DATE'] = pd.to_datetime(df_shooting['OCCUR_DATE'], errors = 'coerce')

In [21]:
#filter for dates that are 2020-01-01 until the present 
df_shooting = df_shooting[df_shooting['OCCUR_DATE'] >= '2020-01-01']

In [22]:
#check the minimum date to make sure filtering worked
df_shooting['OCCUR_DATE'].min()

Timestamp('2020-01-01 00:00:00')

### Drop Duplicates

Duplicates in this case mean multiple victims for one incident, so duplicates will be okay to keep. 

### Make columns to match NYPD Complaint Data

In [23]:
#first check data types of each attribute
df_shooting.dtypes

INCIDENT_KEY                        int64
OCCUR_DATE                 datetime64[ns]
OCCUR_TIME                         object
BORO                               object
PRECINCT                            int64
JURISDICTION_CODE                 float64
LOCATION_DESC                      object
STATISTICAL_MURDER_FLAG              bool
PERP_AGE_GROUP                     object
PERP_SEX                           object
PERP_RACE                          object
VIC_AGE_GROUP                      object
VIC_SEX                            object
VIC_RACE                           object
X_COORD_CD                         object
Y_COORD_CD                         object
Latitude                          float64
Longitude                         float64
Lon_Lat                            object
dtype: object

In [24]:
#rename certain columns to match NYPD complaint data 
df_shooting.rename(columns = {'OCCUR_DATE': 'CMPLNT_FR_DT', 'INCIDENT_KEY':'CMPLNT_NUM', 'STATISTICAL_MURDER_FLAG':'OFNS_DESC'}, inplace = True)

In [25]:
#convert lat and Lon to strings
df_shooting['Latitude'] = df_shooting['Latitude'].astype(str)
df_shooting['Longitude'] = df_shooting['Longitude'].astype(str)

In [26]:
#need to make "Lat_Lon" column
df_shooting['Lat_Lon'] = df_shooting['Latitude'].str.cat(df_shooting['Longitude'], sep = ",")

In [27]:
#insert "()" in the "lat_Lon" column
df_shooting['Lat_Lon'] = ('(' + df_shooting['Lat_Lon'] + ')')

In [28]:
#Label OFNS_DESC column as being either SHOOTING INCIDENT or SHOOTING INCIDENT MURDER
#If it is false it is a SHOOTING INCIDENT, if it is true it is a SHOOTING INCIDENT MURDER

df_shooting['OFNS_DESC'] = df_shooting['OFNS_DESC'].apply(lambda x: 'SHOOTING INCIDENT' if x == False else 'SHOOTING INCIDENT MURDER')

In [29]:
df_shooting.head()

CMPLNT_NUM CMPLNT_FR_DT OCCUR_TIME       BORO  PRECINCT  \
21626   214366724   2020-06-19   01:36:00     QUEENS       103   
21627   215350310   2020-07-15   17:54:00   BROOKLYN        73   
21628   211362502   2020-03-20   20:19:00  MANHATTAN        25   
21629   221490639   2020-12-05   23:35:00     QUEENS       110   
21630   213050797   2020-05-13   20:24:00   BROOKLYN        67   

       JURISDICTION_CODE            LOCATION_DESC                 OFNS_DESC  \
21626                0.0          COMMERCIAL BLDG         SHOOTING INCIDENT   
21627                0.0  MULTI DWELL - APT BUILD         SHOOTING INCIDENT   
21628                0.0                      NaN         SHOOTING INCIDENT   
21629                0.0                      NaN         SHOOTING INCIDENT   
21630                0.0                      NaN  SHOOTING INCIDENT MURDER   

      PERP_AGE_GROUP PERP_SEX PERP_RACE VIC_AGE_GROUP VIC_SEX  \
21626            NaN      NaN       NaN         25-44       M   
21627            NaN      NaN       NaN           <18       M   
21628            NaN      NaN       NaN         25-44       M   
21629            NaN      NaN       NaN           65+       M   
21630            NaN      NaN       NaN         25-44       M   

                       VIC_RACE X_COORD_CD Y_COORD_CD            Latitude  \
21626                     BLACK    1016807     179007  40.657957922000044   
21627                     BLACK    1011811     246833  40.844139945000045   
21628                     BLACK    1009320     244794   40.83855090900005   
21629  ASIAN / PACIFIC ISLANDER    1001350     201326   40.71926178300004   
21630                     BLACK    1051960     160392  40.606664855000076   

                Longitude                                        Lon_Lat  \
21626  -73.88266012399998  POINT (-73.88266012399998 40.657957922000044)   
21627  -73.90038861799998   POINT (-73.90038861799998 40.84413994500005)   
21628  -73.90939937399997   POINT (-73.90939937399997 40.83855090900005)   
21629  -73.93831327799995   POINT (-73.93831327799995 40.71926178300004)   
21630  -73.75614960299998  POINT (-73.75614960299998 40.606664855000076)   

                                       Lat_Lon  
21626  (40.657957922000044,-73.88266012399998)  
21627  (40.844139945000045,-73.90038861799998)  
21628   (40.83855090900005,-73.90939937399997)  
21629   (40.71926178300004,-73.93831327799995)  
21630  (40.606664855000076,-73.75614960299998)

In [30]:
#select columns
df_shooting = df_shooting[['CMPLNT_NUM', 'CMPLNT_FR_DT','OFNS_DESC','Lat_Lon']]

In [31]:
#append df_shooting rows to df_crime
df_crime = df_shooting.append(df_crime)

In [32]:
df_crime

CMPLNT_NUM CMPLNT_FR_DT                       OFNS_DESC  \
21626     214366724   2020-06-19               SHOOTING INCIDENT   
21627     215350310   2020-07-15               SHOOTING INCIDENT   
21628     211362502   2020-03-20               SHOOTING INCIDENT   
21629     221490639   2020-12-05               SHOOTING INCIDENT   
21630     213050797   2020-05-13        SHOOTING INCIDENT MURDER   
...             ...          ...                             ...   
7396612   116305536   2020-01-02                   GRAND LARCENY   
7396613   234726969   2020-01-05  CRIMINAL MISCHIEF & RELATED OF   
7396616   950546587   2020-01-04               CRIMINAL TRESPASS   
7396617   531630504   2020-01-03                  FELONY ASSAULT   
7396618   250138445   2020-01-02                   GRAND LARCENY   

                                         Lat_Lon CMPLNT_FR_TM CMPLNT_TO_DT  \
21626    (40.657957922000044,-73.88266012399998)          NaN          NaN   
21627    (40.844139945000045,-73.90038861799998)          NaN          NaN   
21628     (40.83855090900005,-73.90939937399997)          NaN          NaN   
21629     (40.71926178300004,-73.93831327799995)          NaN          NaN   
21630    (40.606664855000076,-73.75614960299998)          NaN          NaN   
...                                          ...          ...          ...   
7396612  (40.81593767200008, -73.95796004899995)     17:00:00   01/02/2020   
7396613  (40.80404600700007, -73.93662010299995)     05:00:00   01/05/2020   
7396616  (40.81242379100007, -73.94953058799997)     20:30:00   01/04/2020   
7396617  (40.71521116400004, -73.74887699299995)     08:30:00   01/03/2020   
7396618  (40.62934515400008, -73.92799197199997)     15:50:00   01/02/2020   

        CMPLNT_TO_TM  ADDR_PCT_CD      RPT_DT  KY_CD  ...  SUSP_RACE SUSP_SEX  \
21626            NaN          NaN         NaN    NaN  ...        NaN      NaN   
21627            NaN          NaN         NaN    NaN  ...        NaN      NaN   
21628            NaN          NaN         NaN    NaN  ...        NaN      NaN   
21629            NaN          NaN         NaN    NaN  ...        NaN      NaN   
21630            NaN          NaN         NaN    NaN  ...        NaN      NaN   
...              ...          ...         ...    ...  ...        ...      ...   
7396612     17:05:00         26.0  01/02/2020  109.0  ...        NaN      NaN   
7396613     07:50:00         25.0  01/05/2020  351.0  ...    UNKNOWN        U   
7396616     20:34:00         32.0  01/04/2020  352.0  ...      BLACK        M   
7396617     08:45:00        105.0  01/03/2020  106.0  ...    UNKNOWN        U   
7396618     15:55:00         63.0  01/02/2020  109.0  ...      BLACK        F   

        TRANSIT_DISTRICT   Latitude  Longitude               PATROL_BORO  \
21626                NaN        NaN        NaN                       NaN   
21627                NaN        NaN        NaN                       NaN   
21628                NaN        NaN        NaN                       NaN   
21629                NaN        NaN        NaN                       NaN   
21630                NaN        NaN        NaN                       NaN   
...                  ...        ...        ...                       ...   
7396612              NaN  40.815938 -73.957960     PATROL BORO MAN NORTH   
7396613              NaN  40.804046 -73.936620     PATROL BORO MAN NORTH   
7396616              NaN  40.812424 -73.949531     PATROL BORO MAN NORTH   
7396617              NaN  40.715211 -73.748877  PATROL BORO QUEENS SOUTH   
7396618              NaN  40.629345 -73.927992   PATROL BORO BKLYN SOUTH   

        STATION_NAME VIC_AGE_GROUP                  VIC_RACE VIC_SEX  
21626            NaN           NaN                       NaN     NaN  
21627            NaN           NaN                       NaN     NaN  
21628            NaN           NaN                       NaN     NaN  
21629            NaN           NaN                       NaN     NaN  
21630         

### Make a Zipcode & Neighborhood Column

It will be easier to see regional data by zipcode. However, there are only latitude and longitude coordinates. 

Using a KD Tree model we can closely approximate these zipcodes. 

In [33]:
#Reset index 
df_crime.reset_index(inplace=True)

In [34]:
#make tuples of the Lat_Lon column
df_crime['Lat_Lon'] = df_crime['Lat_Lon'].apply(lambda x: eval(str(x)))

In [35]:
#Create location dataframe from census data
#code reference: https://www.codementor.io/@bobhaffner/reverse-geocoding-bljjp5byw
#zip reference: https://www2.census.gov/geo/docs/maps-data/data/gazetteer/2020_Gazetteer/2020_Gaz_zcta_national.zip

#create location dataframe from census data 
df_locations = pd.read_csv('2020_Gaz_zcta_national.zip', dtype={'GEOID':'str'}, sep='\t')

#column cleanup
df_locations.columns = df_locations.columns.str.strip()

#print df_locations to inspect it
print (len(df_locations)) 
print(df_locations.head())

33144
   GEOID      ALAND   AWATER  ALAND_SQMI  AWATER_SQMI   INTPTLAT  INTPTLONG
0  00601  166659744   799292      64.348        0.309  18.180555 -66.749961
1  00602   79307538  4428428      30.621        1.710  18.361945 -67.175597
2  00603   81887203   181412      31.617        0.070  18.455183 -67.119887
3  00606  109579950    12487      42.309        0.005  18.158327 -66.932928
4  00610   93013430  4172059      35.913        1.611  18.294032 -67.127156


In [36]:
#KDT Model
kdt = KDTree(df_locations[['INTPTLAT', 'INTPTLONG']])

In [37]:
#Create loop to make a list of zip codes that correspond to the lat and long in 'lat_Lon' column
list1=[]
for i in range(len(df_crime['Lat_Lon'])):
    nyc_point = np.array((df_crime['Lat_Lon'][i]))
    nyc_point_kdt = np.expand_dims(nyc_point, axis=0)
    list1.append(df_locations.loc[kdt.query(nyc_point_kdt, k=1, return_distance=False)[0], 'GEOID'].values)

In [38]:
#need to "remove axes of length one in numpy array that was created"
#https://numpy.org/doc/stable/reference/generated/numpy.squeeze.html
list1 = np.squeeze(list1)

In [39]:
#Make a new zipcode column
df_crime['zipcode'] = list1

In [40]:
df_crime.head()

index  CMPLNT_NUM CMPLNT_FR_DT                 OFNS_DESC  \
0  21626   214366724   2020-06-19         SHOOTING INCIDENT   
1  21627   215350310   2020-07-15         SHOOTING INCIDENT   
2  21628   211362502   2020-03-20         SHOOTING INCIDENT   
3  21629   221490639   2020-12-05         SHOOTING INCIDENT   
4  21630   213050797   2020-05-13  SHOOTING INCIDENT MURDER   

                                    Lat_Lon CMPLNT_FR_TM CMPLNT_TO_DT  \
0  (40.657957922000044, -73.88266012399998)          NaN          NaN   
1  (40.844139945000045, -73.90038861799998)          NaN          NaN   
2   (40.83855090900005, -73.90939937399997)          NaN          NaN   
3   (40.71926178300004, -73.93831327799995)          NaN          NaN   
4  (40.606664855000076, -73.75614960299998)          NaN          NaN   

  CMPLNT_TO_TM  ADDR_PCT_CD RPT_DT  ...  SUSP_SEX  TRANSIT_DISTRICT Latitude  \
0          NaN          NaN    NaN  ...       NaN               NaN      NaN   
1          NaN          NaN    NaN  ...       NaN               NaN      NaN   
2          NaN          NaN    NaN  ...       NaN               NaN      NaN   
3          NaN          NaN    NaN  ...       NaN               NaN      NaN   
4          NaN          NaN    NaN  ...       NaN               NaN      NaN   

  Longitude PATROL_BORO STATION_NAME VIC_AGE_GROUP VIC_RACE VIC_SEX  zipcode  
0       NaN         NaN          NaN           NaN      NaN     NaN    11239  
1       NaN         NaN          NaN           NaN      NaN     NaN    10457  
2       NaN         NaN          NaN           NaN      NaN     NaN    10456  
3       NaN         NaN          NaN           NaN      NaN     NaN    11222  
4       NaN         NaN          NaN           NaN      NaN     NaN    11691  

[5 rows x 37 columns]

### Modified Zip Code Tabulation Areas (MODZCTA) - Match Zipcode Data with MODZCTA

In regional data it is recommended to represent zipcode data as MODZCTA data, which are basically modified zipcodes that represent neighborhoods. We will download this data from NYC Open Data. 

In [41]:
#Data Source
#https://data.cityofnewyork.us/Health/Modified-Zip-Code-Tabulation-Areas-MODZCTA-/pri4-ifjk/data?no_mobile=true

In [42]:
#read in data to make MODZCTA dataframe
df_modzcta = pd.read_csv('Modified_Zip_Code_Tabulation_Areas__MODZCTA_.csv')

In [43]:
pd.set_option('display.max_rows', 179) #allow us to see all rows
df_modzcta

MODZCTA         label                                               ZCTA  \
0      10001  10001, 10118                                10001, 10119, 10199   
1      10002         10002                                              10002   
2      10003         10003                                              10003   
3      10026         10026                                              10026   
4      10004         10004                                              10004   
5      10005         10005                                       10005, 10271   
6      10006         10006                                              10006   
7      10007         10007                                10007, 10278, 10279   
8      10009         10009                                              10009   
9      10010         10010                                              10010   
10     10011         10011                                              10011   
11     10012         10012                                              10012   
12     10013         10013                                              10013   
13     10014         10014                                              10014   
14     10016         10016                                              10016   
15     10017         10017  10017, 10165, 10167, 10168, 10169, 10170, 1017...   
16     10018         10018                                              10018   
17     10030         10030                                              10030   
18     10019  10019, 10020                  10019, 10020, 10103, 10111, 10112   
19     10021         10021                                              10021   
20     10022         10022                         10022, 10152, 10153, 10154   
21     10023         10023                                              10023   
22     10024         10024                                              10024   
23     10025         10025                                              10025   
24     10027         10027                                       10027, 10115   
25     10028         10028                                              10028   
26     10029         10029                                              10029   
27     10031         10031                                              10031   
28     10032         10032                                              10032   
29     10033         10033                                              10033   
30     10034         10034                                              10034   
31     10037         10037                                              10037   
32     10035         10035                                              10035   
33     10036         10036                                       10036, 10110   
34     10038         10038                                              10038   
35     10040         10040                                              10040   
36     11104         11104                                              11104   
37     11105         11105                                              11105   
38     10039         10039                                              10039   
39     10128         10128                                              10128   
40     10044         10044                                              10044   
41     10065         10065                                              10065   
42     10069         10069                                              10069   
43     10075  10075, 10162                                       10075, 10162   
44     10280         10280                                              10280   
45     10282         10282                                              10282   
46     10301         10301                                              10301   
47     11106         11106                                              11106   
48     10302         10302          

The 'label' column corresponds to zipcodes. 

In [44]:
#remove null values 
df_modzcta.dropna(inplace= True)

In [45]:
df_modzcta

MODZCTA         label                                               ZCTA  \
0      10001  10001, 10118                                10001, 10119, 10199   
1      10002         10002                                              10002   
2      10003         10003                                              10003   
3      10026         10026                                              10026   
4      10004         10004                                              10004   
5      10005         10005                                       10005, 10271   
6      10006         10006                                              10006   
7      10007         10007                                10007, 10278, 10279   
8      10009         10009                                              10009   
9      10010         10010                                              10010   
10     10011         10011                                              10011   
11     10012         10012                                              10012   
12     10013         10013                                              10013   
13     10014         10014                                              10014   
14     10016         10016                                              10016   
15     10017         10017  10017, 10165, 10167, 10168, 10169, 10170, 1017...   
16     10018         10018                                              10018   
17     10030         10030                                              10030   
18     10019  10019, 10020                  10019, 10020, 10103, 10111, 10112   
19     10021         10021                                              10021   
20     10022         10022                         10022, 10152, 10153, 10154   
21     10023         10023                                              10023   
22     10024         10024                                              10024   
23     10025         10025                                              10025   
24     10027         10027                                       10027, 10115   
25     10028         10028                                              10028   
26     10029         10029                                              10029   
27     10031         10031                                              10031   
28     10032         10032                                              10032   
29     10033         10033                                              10033   
30     10034         10034                                              10034   
31     10037         10037                                              10037   
32     10035         10035                                              10035   
33     10036         10036                                       10036, 10110   
34     10038         10038                                              10038   
35     10040         10040                                              10040   
36     11104         11104                                              11104   
37     11105         11105                                              11105   
38     10039         10039                                              10039   
39     10128         10128                                              10128   
40     10044         10044                                              10044   
41     10065         10065                                              10065   
42     10069         10069                                              10069   
43     10075  10075, 10162                                       10075, 10162   
44     10280         10280                                              10280   
45     10282         10282                                              10282   
46     10301         10301                                              10301   
47     11106         11106                                              11106   
48     10302         10302          

In [46]:
#select only the MODZCTA, label and pop_est columns
df_modzcta = df_modzcta[['MODZCTA', 'label', 'pop_est']]

In [47]:
df_modzcta

MODZCTA         label  pop_est
0      10001  10001, 10118    23072
1      10002         10002    74993
2      10003         10003    54682
3      10026         10026    39363
4      10004         10004     3028
5      10005         10005     8831
6      10006         10006     3454
7      10007         10007     7023
8      10009         10009    57925
9      10010         10010    33730
10     10011         10011    50472
11     10012         10012    23411
12     10013         10013    28709
13     10014         10014    30738
14     10016         10016    51057
15     10017         10017    15801
16     10018         10018     9687
17     10030         10030    29882
18     10019  10019, 10020    43703
19     10021         10021    46215
20     10022         10022    31125
21     10023         10023    62435
22     10024         10024    59001
23     10025         10025    92805
24     10027         10027    64548
25     10028         10028    47793
26     10029         10029    79597
27     10031         10031    60254
28     10032         10032    64264
29     10033         10033    59607
30     10034         10034    42910
31     10037         10037    21139
32     10035         10035    35743
33     10036         10036    27428
34     10038         10038    22800
35     10040         10040    44571
36     11104         11104    26119
37     11105         11105    37382
38     10039         10039    27726
39     10128         10128    59256
40     10044         10044    11682
41     10065         10065    28109
42     10069         10069     5085
43     10075  10075, 10162    23140
44     10280         10280     9384
45     10282         10282     5593
46     10301         10301    38733
47     11106         11106    38363
48     10302         10302    18204
49     10303         10303    25947
50     10304         10304    41064
51     10451         10451    47798
52     10305         10305    42298
53     10306         10306    53142
54     10307         10307    15232
55     10308         10308    30186
56     10309         10309    33531
57     10310         10310    23262
58     10312         10312    61741
59     10314         10314    90761
60     10456         10456    94218
61     10452         10452    76078
62     10453         10453    79606
63     10454         10454    38203
64     10455         10455    42283
65     10457         10457    74554
66     10458         10458    83960
67     10459         10459    48351
68     10460         10460    59505
69     10461         10461    50348
70     11004  11004, 11005    19216
71     11101         11101    30043
72     10462         10462    76756
73     10463         10463    71360
74     10464         10464     4434
75     10465         10465    44205
76     10466         10466    72479
77     10467         10467   103732
78     10468         10468    78647
79     10469         10469    73870
80     10470         10470    15432
81     11102         11102    30800
82     11109         11109     5981
83     10471         10471    21896
84     10472         10472    68796
85     10473         10473    59609
86     10474         10474    12127
87     11103         11103    36140
88     10475         10475    43208
89     11201         11201    62823
90     11203         11203    75451
91     11204         11204    76978
92     11205         11205    46064
93     11206         11206    88349
94     11207         11207    91972
95     11209         11209    67809
96     11208         11208    98660
97     11210         11210    67040
98     11211  11211, 11249   102624
99     11212         11212    76527
100    11213         11213    66295
101    11218         11218    74508
102    11214         11214    92946
103    11215         11215    70156
104    11216         11216    56308
105    11217  11217, 11243    41407
106    11219         11219    90036
107    11220         11220    95643
108    11221         11221    83835
109    11222         11222    36492
110    11

In [48]:
#Need to convert fields with multiple data into multiple rows
#Code reference: https://sureshssarda.medium.com/pandas-splitting-exploding-a-column-into-multiple-rows-b1b1d59ea12e
new_df = pd.DataFrame(df_modzcta['label'].str.split(',').tolist(), index=df_modzcta['MODZCTA']).stack()
new_df = new_df.reset_index([0, 'MODZCTA'])
new_df.columns = ['MODZCTA', 'zipcode']
new_df

MODZCTA zipcode
0      10001   10001
1      10001   10118
2      10002   10002
3      10003   10003
4      10026   10026
..       ...     ...
178    11433   11433
179    11434   11434
180    11435   11435
181    11694   11694
182    11697   11697

[183 rows x 2 columns]

In [49]:
#merge new_df and df_modzcta using a left join
df_modzcta = new_df.merge(df_modzcta, how = 'left', on ='MODZCTA')

In [50]:
#select columns
df_modzcta = df_modzcta[['MODZCTA', 'zipcode','pop_est']]

In [51]:
df_modzcta.head()

MODZCTA zipcode  pop_est
0    10001   10001    23072
1    10001   10118    23072
2    10002   10002    74993
3    10003   10003    54682
4    10026   10026    39363

### Neighborhood Names - Match Neighborhood Names to MODZCTA

It is also useful to match neighborhood names to MODZCTA data. So we will use the NYC Cvodi-19 dataset which contains this. 

In [52]:
#Data Source
#https://raw.githubusercontent.com/nychealth/coronavirus-data/master/latest/last7days-by-modzcta.csv

In [53]:
#import data to make Neighborhood Name Dataframe
df_names = pd.read_csv('https://raw.githubusercontent.com/nychealth/coronavirus-data/master/latest/last7days-by-modzcta.csv')

In [54]:
df_names.head()

modzcta                             modzcta_name         label        lat  \
0    10001               Chelsea/NoMad/West Chelsea  10001, 10118  40.750693   
1    10002                Chinatown/Lower East Side         10002  40.715781   
2    10003  East Village/Gramercy/Greenwich Village         10003  40.731825   
3    10004                       Financial District         10004  40.703675   
4    10005                       Financial District         10005  40.706092   

         lon  percentpositivity_7day  people_tested  people_positive  \
0 -73.997137                    0.87           1036                9   
1 -73.986176                    1.66           2343               39   
2 -73.989164                    0.96           3220               31   
3 -74.013106                    0.94            106                1   
4 -74.008861                    0.72            276                2   

   median_daily_test_rate adequately_tested       daterange  
0                   738.8               Yes  April 29-May 5  
1                   543.0               Yes  April 29-May 5  
2                  1106.0               Yes  April 29-May 5  
3                   403.8               Yes  April 29-May 5  
4                   502.4               Yes  April 29-May 5

In [55]:
#select columns 
df_names = df_names[['modzcta','modzcta_name','label']]

In [56]:
df_names.head()

modzcta                             modzcta_name         label
0    10001               Chelsea/NoMad/West Chelsea  10001, 10118
1    10002                Chinatown/Lower East Side         10002
2    10003  East Village/Gramercy/Greenwich Village         10003
3    10004                       Financial District         10004
4    10005                       Financial District         10005

In [57]:
#rename modzcta column
df_names.columns = ['MODZCTA','NeighborhoodName','ZipsThatMakeUpMODZCTA']

In [58]:
#merge df_names and df_modzcta into new dataframe called df_nyclocation
df_nyclocation = df_modzcta.merge(df_names, how = 'left', on ='MODZCTA')

In [59]:
df_nyclocation.head()

MODZCTA zipcode  pop_est                         NeighborhoodName  \
0    10001   10001    23072               Chelsea/NoMad/West Chelsea   
1    10001   10118    23072               Chelsea/NoMad/West Chelsea   
2    10002   10002    74993                Chinatown/Lower East Side   
3    10003   10003    54682  East Village/Gramercy/Greenwich Village   
4    10026   10026    39363                   Central Harlem (South)   

  ZipsThatMakeUpMODZCTA  
0          10001, 10118  
1          10001, 10118  
2                 10002  
3                 10003  
4                 10026

In [60]:
#check column types, for MODZCTA we want it to be a string not an integer
df_nyclocation.dtypes

MODZCTA                   int64
zipcode                  object
pop_est                   int64
NeighborhoodName         object
ZipsThatMakeUpMODZCTA    object
dtype: object

In [61]:
#convert MODZCTA column into strings
df_nyclocation['MODZCTA'] = df_nyclocation['MODZCTA'].astype(str)

### Merge NYC Location Data to Crime Data

In [62]:
#merge df_nyclocation with df_crime
df_crime = df_crime.merge(df_nyclocation, how = 'left', on = 'zipcode')

In [63]:
#search for missing values in 'NeighborhoodName': nulls in this column means that these locations did not occur in a specific neighborhood but maybe in a park or airport.
#search for missing values in 'OFNS_DESC': we will sort by crime description so null values are not very specific and should be removed. 
df_crime.isnull().sum()

index                         0
CMPLNT_NUM                    0
CMPLNT_FR_DT                  0
OFNS_DESC                     5
Lat_Lon                       0
CMPLNT_FR_TM               1942
CMPLNT_TO_DT              39478
CMPLNT_TO_TM              39359
ADDR_PCT_CD                1942
RPT_DT                     1942
KY_CD                      1942
PD_CD                      1942
PD_DESC                    1942
CRM_ATPT_CPTD_CD           1942
LAW_CAT_CD                 1942
BORO_NM                    1942
LOC_OF_OCCUR_DESC         67202
PREM_TYP_DESC              2605
JURIS_DESC                 1942
JURISDICTION_CODE          1942
PARKS_NM                 403718
HADEVELOPT               404834
HOUSING_PSA              375954
X_COORD_CD                 1942
Y_COORD_CD                 1942
SUSP_AGE_GROUP            93841
SUSP_RACE                 93841
SUSP_SEX                  93841
TRANSIT_DISTRICT         399429
Latitude                   1942
Longitude                  1942
PATROL_B

In [64]:
#remove nulls from 'NeighborhoodName' column
df_crime = df_crime[pd.notnull(df_crime['NeighborhoodName'])]

In [65]:
#remove nulls from 'OFNS_DESC ' column
df_crime = df_crime[pd.notnull(df_crime['OFNS_DESC'])]

In [66]:
df_crime.isnull().sum()

index                         0
CMPLNT_NUM                    0
CMPLNT_FR_DT                  0
OFNS_DESC                     0
Lat_Lon                       0
CMPLNT_FR_TM               1900
CMPLNT_TO_DT              37373
CMPLNT_TO_TM              37259
ADDR_PCT_CD                1900
RPT_DT                     1900
KY_CD                      1900
PD_CD                      1900
PD_DESC                    1900
CRM_ATPT_CPTD_CD           1900
LAW_CAT_CD                 1900
BORO_NM                    1900
LOC_OF_OCCUR_DESC         64003
PREM_TYP_DESC              2467
JURIS_DESC                 1900
JURISDICTION_CODE          1900
PARKS_NM                 386548
HADEVELOPT               387654
HOUSING_PSA              359113
X_COORD_CD                 1900
Y_COORD_CD                 1900
SUSP_AGE_GROUP            90020
SUSP_RACE                 90020
SUSP_SEX                  90020
TRANSIT_DISTRICT         383148
Latitude                   1900
Longitude                  1900
PATROL_B

In [67]:
df_crime.head()

index  CMPLNT_NUM CMPLNT_FR_DT                 OFNS_DESC  \
0  21626   214366724   2020-06-19         SHOOTING INCIDENT   
1  21627   215350310   2020-07-15         SHOOTING INCIDENT   
2  21628   211362502   2020-03-20         SHOOTING INCIDENT   
3  21629   221490639   2020-12-05         SHOOTING INCIDENT   
4  21630   213050797   2020-05-13  SHOOTING INCIDENT MURDER   

                                    Lat_Lon CMPLNT_FR_TM CMPLNT_TO_DT  \
0  (40.657957922000044, -73.88266012399998)          NaN          NaN   
1  (40.844139945000045, -73.90038861799998)          NaN          NaN   
2   (40.83855090900005, -73.90939937399997)          NaN          NaN   
3   (40.71926178300004, -73.93831327799995)          NaN          NaN   
4  (40.606664855000076, -73.75614960299998)          NaN          NaN   

  CMPLNT_TO_TM  ADDR_PCT_CD RPT_DT  ...  PATROL_BORO  STATION_NAME  \
0          NaN          NaN    NaN  ...          NaN           NaN   
1          NaN          NaN    NaN  ...          NaN           NaN   
2          NaN          NaN    NaN  ...          NaN           NaN   
3          NaN          NaN    NaN  ...          NaN           NaN   
4          NaN          NaN    NaN  ...          NaN           NaN   

  VIC_AGE_GROUP VIC_RACE VIC_SEX zipcode MODZCTA  pop_est  \
0           NaN      NaN     NaN   11239   11239  13244.0   
1           NaN      NaN     NaN   10457   10457  74554.0   
2           NaN      NaN     NaN   10456   10456  94218.0   
3           NaN      NaN     NaN   11222   11222  36492.0   
4           NaN      NaN     NaN   11691   11691  67094.0   

                       NeighborhoodName  ZipsThatMakeUpMODZCTA  
0                         East New York                  11239  
1  Belmont/Claremont/Mount Hope/Tremont                  10457  
2                  Claremont/Morrisania                  10456  
3                            Greenpoint                  11222  
4                 Edgemere/Far Rockaway                  11691  

[5 rows x 41 columns]

All crimes in df_crime were committed in New York City

### Clean Dataset Description: NYPD_Complaint_Data_Historic_Clean.csv

- The dataset was filtered to include non-missing 2020 NYPD crime data (including shootings & murders) in NYC only. 
- There was a focus on 2020 due to the large size of the original data and it being the most recent. 
- Location and neighborhood data were added for each crime:
    - zipcode	
    - MODZCTA	
    - pop_est	
    - NeighborhoodName	
    - ZipsThatMakeUpMODZCTA

In [68]:
df_crime.shape

(389074, 41)

### Export Clean Dataset

In [69]:
df_crime.to_csv("NYPD_Complaint_Data_Historic_Clean.csv")